Pyrado/pyrado/algorithms/meta/pddr.py

```diff
@@ -0,0 +1,327 @@
+
+            # Update policy and value function
+            self.update(rollouts=ros)
+
+    def sample(self):
+        """ Samples observations from several samplers. """
+        ros = []
+        rets = []
+        all_lengths = []
+        for sampler in self.samplers:
+            samples = sampler.sample()
+            ros.append(samples)
+            rets.extend([sample.undiscounted_return() for sample in samples])
+            all_lengths.extend([sample.length for sample in samples])
+
+        return ros, np.array(rets), np.array(all_lengths)
+
+    def update(self, *args: Any, **kwargs: Any):
+        """ Update the policy's (and value functions') parameters based on the collected rollout data. """
+        obss = []
+        losses = []
+        for t in range(self.num_teachers):
+            concat_ros = StepSequence.concat(kwargs["rollouts"][t])
+            concat_ros.torch(data_type=to.get_default_dtype())
+            obss.append(concat_ros.get_data_values("observations")[: self.min_steps])
+
+        # Train student
+        for epoch in range(self.num_epochs):
+            self.optimizer.zero_grad()
+
+            loss = 0
+            for t_idx, teacher in enumerate(self.teacher_policies):
+                s_dist = self.expl_strat.action_dist_at(self.policy(obss[t_idx]))
+                s_act = s_dist.sample()
+                t_dist = self.teacher_expl_strat[t_idx].action_dist_at(teacher(obss[t_idx]))
+
+                l = self.teacher_weights[t_idx] * self.criterion(t_dist.log_prob(s_act), s_dist.log_prob(s_act))
+                loss += l
+                losses.append([t_idx, l.item()])
+            print(f"Epoch {epoch} Loss: {loss.item()}")
+            loss.backward()
+            self.optimizer.step()
+
+    def save_snapshot(self, meta_info: dict = None):
+        super().save_snapshot(meta_info)
+
+        pyrado.save(self.policy, "policy.pt", self.save_dir)
+
+        if meta_info is None:
+            # This algorithm instance is not a subroutine of another algorithm
+            pyrado.save(self.env_real, "env.pkl", self.save_dir)
+
+    def _train_teacher(self, idx: int, snapshot_mode: str = "latest", seed: int = None):
+        """ Wrapper for use of multiprocessing: Trains oe teacher. """
+        self.algos[idx].train(snapshot_mode, seed, meta_info={prefix: f"teacher_{idx}"})
+
+    def set_random_envs(self):
+        """ Creates random environments of the given type. """
+        self.randomizer.randomize(num_samples=self.num_teachers)
+        params = self.randomizer.get_params(fmt="dict", dtype="numpy")
+
+        for e in range(self.num_teachers):
+            self.teacher_envs.append(deepcopy(self.env_real))
+            print({key: value[e] for key, value in params.items()})
+            self.teacher_envs[e].domain_param = {key: value[e] for key, value in params.items()}
+
+    def train_teachers(self, snapshot_mode: str = "latest", seed: int = None):
+        for idx in range(self.num_teachers):
```

Comment:
I think it would be nice to add a print inside the for loop here here (see SPOTA), sth like:
print_cbt(f"Training teacher {idx + 1} of {self.num_teachers}... ", "c")